File: web-client/src/services/responsiveLayoutAdapter.ts
/**
 * Responsive Layout Adapter
 *
 * Handles window size-based layout adjustments and dynamic layout changes.
 * Implements requirement 12.5 for document table parts.
 */

import { ref, computed, onUnmounted, type Ref } from 'vue'
import type { FormField, LayoutConfiguration } from './formLayoutManager'
import { FormLayoutManager } from './formLayoutManager'

export interface WindowSize {
  width: number
  height: number
}

export interface ResponsiveLayoutRules {
  minWidthForTwoColumns: number
  fieldWidthRatios: Record<string, number>
  breakpoints: Record<string, number>
}

export class ResponsiveLayoutAdapter {
  private rules: ResponsiveLayoutRules
  private layoutManager: FormLayoutManager
  private currentWindowSize: WindowSize | null = null
  private currentLayoutType: 'single_column' | 'two_column' = 'single_column'
  private fields: FormField[] = []
  private resizeTimeout: number | null = null
  private readonly resizeDebounceMs = 150

  // Reactive properties
  public windowSize: Ref<WindowSize | null> = ref(null)
  public layoutType: Ref<'single_column' | 'two_column'> = ref('single_column')
  public breakpoint: Ref<string> = ref('desktop')

  constructor() {
    this.rules = {
      minWidthForTwoColumns: 800,
      fieldWidthRatios: {
        short: 0.3,
        medium: 0.5,
        long: 1.0,
        reference: 0.4
      },
      breakpoints: {
        mobile: 480,
        tablet: 768,
        desktop: 1024,
        wide: 1440
      }
    }

    this.layoutManager = new FormLayoutManager()

    // Set up window resize listener
    this.setupResizeListener()
  }

  private resizeHandler: (() => void) | null = null

  private setupResizeListener(): void {
    this.resizeHandler = () => {
      this.handleWindowResize({
        width: window.innerWidth,
        height: window.innerHeight
      })
    }

    window.addEventListener('resize', this.resizeHandler)
    
    // Initial size
    this.resizeHandler()
  }

  cleanup(): void {
    if (this.resizeHandler) {
      window.removeEventListener('resize', this.resizeHandler)
    }
    if (this.resizeTimeout) {
      clearTimeout(this.resizeTimeout)
    }
  }

  setFields(fields: FormField[]): void {
    this.fields = fields
  }

  handleWindowResize(newSize: WindowSize): void {
    this.currentWindowSize = newSize
    this.windowSize.value = newSize

    // Update breakpoint
    this.breakpoint.value = this.getCurrentBreakpoint()

    // Debounce resize events
    if (this.resizeTimeout) {
      clearTimeout(this.resizeTimeout)
    }

    this.resizeTimeout = window.setTimeout(() => {
      this.handleResizeTimeout()
    }, this.resizeDebounceMs)
  }

  private handleResizeTimeout(): void {
    if (!this.currentWindowSize || !this.fields) {
      return
    }

    // Determine optimal layout
    const shouldUseTwoColumns = this.shouldUseTwoColumns(
      this.currentWindowSize,
      this.fields.filter((f) => this.layoutManager.isShortField(f)).length
    )

    const newLayoutType: 'single_column' | 'two_column' = shouldUseTwoColumns
      ? 'two_column'
      : 'single_column'

    // Update layout type if changed
    if (newLayoutType !== this.currentLayoutType) {
      this.currentLayoutType = newLayoutType
      this.layoutType.value = newLayoutType
    }
  }

  shouldUseTwoColumns(windowSize: WindowSize, fieldCount: number): boolean {
    return windowSize.width >= this.rules.minWidthForTwoColumns && fieldCount >= 6
  }

  adaptToWindowSize(layout: LayoutConfiguration, windowSize: WindowSize): LayoutConfiguration {
    const adaptedLayout: LayoutConfiguration = {
      leftColumn: [],
      rightColumn: [],
      fullWidthFields: [...layout.fullWidthFields],
      columnRatio: this.getColumnRatio(windowSize)
    }

    // Redistribute fields based on new ratio
    const allShortFields = [...layout.leftColumn, ...layout.rightColumn]
    if (allShortFields.length > 0) {
      const splitPoint = Math.floor(allShortFields.length * adaptedLayout.columnRatio)
      adaptedLayout.leftColumn = allShortFields.slice(0, splitPoint)
      adaptedLayout.rightColumn = allShortFields.slice(splitPoint)
    }

    return adaptedLayout
  }

  getColumnRatio(windowSize: WindowSize): number {
    if (windowSize.width < this.rules.breakpoints.tablet) {
      return 0.5 // Equal columns on small screens
    } else if (windowSize.width < this.rules.breakpoints.desktop) {
      return 0.45 // Slightly favor right column
    } else {
      return 0.4 // More space for right column on large screens
    }
  }

  getCurrentBreakpoint(): string {
    if (!this.currentWindowSize) {
      return 'desktop'
    }

    const width = this.currentWindowSize.width

    if (width < this.rules.breakpoints.mobile) {
      return 'mobile'
    } else if (width < this.rules.breakpoints.tablet) {
      return 'tablet'
    } else if (width < this.rules.breakpoints.wide) {
      return 'desktop'
    } else {
      return 'wide'
    }
  }

  shouldForceSingleColumn(): boolean {
    if (!this.currentWindowSize) {
      return false
    }

    return this.currentWindowSize.width < this.rules.minWidthForTwoColumns
  }

  getFieldWidthForBreakpoint(fieldType: string): number {
    const breakpoint = this.getCurrentBreakpoint()
    const baseRatio = this.rules.fieldWidthRatios[fieldType] || 0.5

    // Adjust based on breakpoint
    if (breakpoint === 'mobile') {
      return Math.min(baseRatio * 1.2, 1.0) // Wider fields on mobile
    } else if (breakpoint === 'tablet') {
      return baseRatio
    } else {
      return baseRatio * 0.9 // Slightly narrower on desktop
    }
  }
}

/**
 * Vue composable for responsive form layout
 */
export function useResponsiveFormLayout(fields: Ref<FormField[]>) {
  const adapter = new ResponsiveLayoutAdapter()
  const layoutManager = new FormLayoutManager()

  // Cleanup on unmount
  onUnmounted(() => {
    adapter.cleanup()
  })

  // Reactive layout configuration
  const layoutConfig = computed(() => {
    if (!fields.value.length) {
      return null
    }

    adapter.setFields(fields.value)
    const analysis = layoutManager.analyzeFields(fields.value)

    if (analysis.recommendedLayout === 'two_column' && !adapter.shouldForceSingleColumn()) {
      const config = layoutManager.createTwoColumnLayout(fields.value)
      return adapter.windowSize.value ? adapter.adaptToWindowSize(config, adapter.windowSize.value) : config
    }

    return null
  })

  // CSS classes for responsive behavior
  const containerClasses = computed(() => {
    const classes = ['form-container']
    
    if (adapter.layoutType.value === 'two_column') {
      classes.push('form-container--two-column')
    } else {
      classes.push('form-container--single-column')
    }

    classes.push(`form-container--${adapter.breakpoint.value}`)
    
    return classes
  })

  return {
    layoutConfig,
    layoutType: adapter.layoutType,
    windowSize: adapter.windowSize,
    breakpoint: adapter.breakpoint,
    containerClasses,
    shouldUseTwoColumns: (fieldCount: number) => {
      return adapter.windowSize.value
        ? adapter.shouldUseTwoColumns(adapter.windowSize.value, fieldCount)
        : false
    },
    getFieldWidthForBreakpoint: adapter.getFieldWidthForBreakpoint.bind(adapter)
  }
}

File: src/services/responsive_layout_adapter.py
"""
Responsive Layout Adapter

Handles window size-based layout adjustments and dynamic layout changes.
Implements requirement 12.5 for document table parts.
"""
from typing import Dict, Any, Optional, Callable
from PyQt6.QtWidgets import QWidget, QLayout
from PyQt6.QtCore import QObject, QSize, QTimer, pyqtSignal
from .form_layout_manager import FormLayoutManager, FormField, LayoutConfiguration


class WindowSize:
    """Window size classification"""
    
    def __init__(self, width: int, height: int):
        self.width = width
        self.height = height
    
    def is_narrow(self, threshold: int = 800) -> bool:
        """Check if window is narrow"""
        return self.width < threshold
    
    def is_wide(self, threshold: int = 1200) -> bool:
        """Check if window is wide"""
        return self.width >= threshold
    
    def __repr__(self) -> str:
        return f"WindowSize({self.width}x{self.height})"


class ResponsiveLayoutRules:
    """Rules for responsive layout behavior"""
    
    def __init__(self):
        # Minimum window width for 2-column layout
        self.min_width_for_two_columns: int = 800
        
        # Field width ratios for different types
        self.field_width_ratios: Dict[str, float] = {
            'short': 0.3,
            'medium': 0.5,
            'long': 1.0,
            'reference': 0.4
        }
        
        # Breakpoints for layout adaptation
        self.breakpoints: Dict[str, int] = {
            'mobile': 480,
            'tablet': 768,
            'desktop': 1024,
            'wide': 1440
        }
    
    def should_use_two_columns(self, window_size: WindowSize, field_count: int) -> bool:
        """Determine if two-column layout should be used"""
        return (
            window_size.width >= self.min_width_for_two_columns and
            field_count >= 6
        )
    
    def get_column_ratio(self, window_size: WindowSize) -> float:
        """Get optimal column ratio based on window size"""
        if window_size.width < self.breakpoints['tablet']:
            return 0.5  # Equal columns on small screens
        elif window_size.width < self.breakpoints['desktop']:
            return 0.45  # Slightly favor right column
        else:
            return 0.4  # More space for right column on large screens


class ResponsiveLayoutAdapter(QObject):
    """
    Handles responsive layout adaptation based on window size changes.
    
    Features:
    - Monitors window size changes
    - Adapts layout based on breakpoints
    - Handles dynamic layout transitions
    - Maintains field distribution optimization
    """
    
    # Signal emitted when layout should be updated
    layout_update_requested = pyqtSignal(str)  # layout_type: "single_column" or "two_column"
    
    def __init__(self, parent: Optional[QObject] = None):
        super().__init__(parent)
        self.rules = ResponsiveLayoutRules()
        self.layout_manager = FormLayoutManager()
        self.current_window_size: Optional[WindowSize] = None
        self.current_layout_type: str = "single_column"
        self.fields: list[FormField] = []
        
        # Debounce timer for resize events
        self.resize_timer = QTimer()
        self.resize_timer.setSingleShot(True)
        self.resize_timer.timeout.connect(self._handle_resize_timeout)
        self.resize_debounce_ms = 150
    
    def set_fields(self, fields: list[FormField]) -> None:
        """Set the fields to be managed"""
        self.fields = fields
    
    def handle_window_resize(self, new_size: QSize) -> None:
        """
        Handle window resize event with debouncing.
        
        Args:
            new_size: New window size
        """
        self.current_window_size = WindowSize(new_size.width(), new_size.height())
        
        # Debounce resize events
        self.resize_timer.start(self.resize_debounce_ms)
    
    def _handle_resize_timeout(self) -> None:
        """Handle resize timeout - perform actual layout adaptation"""
        if not self.current_window_size or not self.fields:
            return
        
        # Determine optimal layout
        should_use_two_columns = self.rules.should_use_two_columns(
            self.current_window_size,
            len([f for f in self.fields if f.is_short_field()])
        )
        
        new_layout_type = "two_column" if should_use_two_columns else "single_column"
        
        # Only emit signal if layout type changed
        if new_layout_type != self.current_layout_type:
            self.current_layout_type = new_layout_type
            self.layout_update_requested.emit(new_layout_type)
    
    def adapt_to_window_size(
        self,
        layout: LayoutConfiguration,
        window_size: WindowSize
    ) -> LayoutConfiguration:
        """
        Adapt layout configuration to window size.
        
        Args:
            layout: Current layout configuration
            window_size: Current window size
            
        Returns:
            Adapted layout configuration
        """
        adapted_layout = LayoutConfiguration()
        adapted_layout.full_width_fields = layout.full_width_fields.copy()
        
        # Get optimal column ratio
        optimal_ratio = self.rules.get_column_ratio(window_size)
        adapted_layout.column_ratio = optimal_ratio
        
        # Redistribute fields based on new ratio
        all_short_fields = layout.left_column + layout.right_column
        if all_short_fields:
            split_point = int(len(all_short_fields) * optimal_ratio)
            adapted_layout.left_column = all_short_fields[:split_point]
            adapted_layout.right_column = all_short_fields[split_point:]
        
        return adapted_layout
    
    def get_current_breakpoint(self) -> str:
        """Get current breakpoint name"""
        if not self.current_window_size:
            return "desktop"
        
        width = self.current_window_size.width
        
        if width < self.rules.breakpoints['mobile']:
            return "mobile"
        elif width < self.rules.breakpoints['tablet']:
            return "tablet"
        elif width < self.rules.breakpoints['wide']:
            return "desktop"
        else:
            return "wide"
    
    def should_force_single_column(self) -> bool:
        """Check if single column should be forced due to narrow window"""
        if not self.current_window_size:
            return False
        
        return self.current_window_size.width < self.rules.min_width_for_two_columns
    
    def get_field_width_for_breakpoint(self, field_type: str) -> float:
        """Get optimal field width ratio for current breakpoint"""
        breakpoint = self.get_current_breakpoint()
        base_ratio = self.rules.field_width_ratios.get(field_type, 0.5)
        
        # Adjust based on breakpoint
        if breakpoint == "mobile":
            return min(base_ratio * 1.2, 1.0)  # Wider fields on mobile
        elif breakpoint == "tablet":
            return base_ratio
        else:
            return base_ratio * 0.9  # Slightly narrower on desktop


class ResponsiveFormWidget(QWidget):
    """
    Widget that automatically adapts its form layout based on window size.
    
    This is a convenience widget that combines FormLayoutManager and ResponsiveLayoutAdapter
    to provide automatic responsive behavior.
    """
    
    def __init__(self, parent: Optional[QWidget] = None):
        super().__init__(parent)
        self.layout_manager = FormLayoutManager()
        self.adapter = ResponsiveLayoutAdapter(self)
        self.fields: list[FormField] = []
        self.current_layout: Optional[QLayout] = None
        
        # Connect adapter signals
        self.adapter.layout_update_requested.connect(self._rebuild_layout)
    
    def set_fields(self, fields: list[FormField]) -> None:
        """Set form fields and build initial layout"""
        self.fields = fields
        self.adapter.set_fields(fields)
        self._rebuild_layout()
    
    def add_field(self, field: FormField) -> None:
        """Add a single field and rebuild layout"""
        self.fields.append(field)
        self.adapter.set_fields(self.fields)
        self._rebuild_layout()
    
    def remove_field(self, field_name: str) -> None:
        """Remove field by name and rebuild layout"""
        self.fields = [f for f in self.fields if f.name != field_name]
        self.adapter.set_fields(self.fields)
        self._rebuild_layout()
    
    def _rebuild_layout(self, layout_type: Optional[str] = None) -> None:
        """Rebuild the form layout"""
        if not self.fields:
            return
        
        # Clear existing layout
        if self.current_layout:
            self._clear_layout(self.current_layout)
        
        # Create new layout
        new_layout = self.layout_manager.create_form_layout(
            self,
            self.fields,
            force_layout=layout_type
        )
        
        self.setLayout(new_layout)
        self.current_layout = new_layout
    
    def _clear_layout(self, layout: QLayout) -> None:
        """Clear all items from layout"""
        while layout.count():
            child = layout.takeAt(0)
            if child.widget():
                child.widget().setParent(None)
            elif child.layout():
                self._clear_layout(child.layout())
    
    def resizeEvent(self, event) -> None:
        """Handle resize events"""
        super().resizeEvent(event)
        self.adapter.handle_window_resize(event.size())
    
    def force_layout_type(self, layout_type: str) -> None:
        """Force specific layout type"""
        self._rebuild_layout(layout_type)